### PR TITLE
Sampling for allocation events

### PR DESCRIPTION
**Pull request by laurit**
Adds an option to sample only every Nth allocation event.

### CODE REVIEW

**Review by theletterf — APPROVED** [review]

**Review by theletterf — COMMENTED** [review]
@laurit I suggest to move the setting right above the deprecated one.

File: profiler/src/main/java/com/splunk/opentelemetry/profiler/allocationsampler/SystematicAllocationEventSampler.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    counter = 0;
+    return true;
+  }
+
+  @Override
+  public void addAttributes(AttributesBuilder builder) {
+    builder.put(SAMPLER_NAME_KEY, "Systematic sampler");
+    builder.put(SAMPLER_INTERVAL_KEY, samplingInterval);
+  }
+}
```
Comment (breedx-splk) [inline profiler/src/main/java/com/splunk/opentelemetry/profiler/allocationsampler/SystematicAllocationEventSampler.java:55]: Please add unit test coverage for this class.

File: profiler/src/main/java/com/splunk/opentelemetry/profiler/allocationsampler/AllocationEventSampler.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+package com.splunk.opentelemetry.profiler.allocationsampler;
+
+import io.opentelemetry.api.common.AttributesBuilder;
+import jdk.jfr.consumer.RecordedEvent;
+
+/** A sampler that decides whether given allocation event should be sampled or discarded. */
+public interface AllocationEventSampler {
+
+  /**
+   * @param event allocation event
```
Comment (breedx-splk) [inline profiler/src/main/java/com/splunk/opentelemetry/profiler/allocationsampler/AllocationEventSampler.java:26]: I think it's worth calling out explicitly which event types implementations may receive.

**Review by breedx-splk — CHANGES_REQUESTED** [review]
This is cool, a few comments. Thanks.

**Review by theletterf — APPROVED** [review]

File: profiler/src/main/java/com/splunk/opentelemetry/profiler/allocationsampler/AlwaysOnAllocationEventSampler.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.splunk.opentelemetry.profiler.allocationsampler;
+
+import io.opentelemetry.api.common.AttributesBuilder;
+import jdk.jfr.consumer.RecordedEvent;
+
+class AlwaysOnAllocationEventSampler implements AllocationEventSampler {
```
[the remark this answers is not in the document]
Comment (laurit) [reply profiler/src/main/java/com/splunk/opentelemetry/profiler/allocationsampler/AlwaysOnAllocationEventSampler.java:22]: This name matches sdk class `AlwaysOnSampler` which is used for sampling traces. Anyway this class is not really needed I only added it to match what sdk uses for tracing. Might as well allow null sampler.

File: profiler/src/main/java/com/splunk/opentelemetry/profiler/allocationsampler/AlwaysOnAllocationEventSampler.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  static AllocationEventSampler INSTANCE = new AlwaysOnAllocationEventSampler();
+
+  @Override
+  public boolean shouldSample(RecordedEvent event) {
+    return true;
+  }
+
+  @Override
+  public void addAttributes(AttributesBuilder builder) {
```
[the remark this answers is not in the document]
Comment (laurit) [reply profiler/src/main/java/com/splunk/opentelemetry/profiler/allocationsampler/AlwaysOnAllocationEventSampler.java:31]: With current implementation this is not desirable. `AlwaysOnAllocationEventSampler` is the default sampler, no need to add these attributes when not really using a sampler.

File: profiler/src/main/java/com/splunk/opentelemetry/profiler/allocationsampler/SystematicAllocationEventSampler.java (hunk 1 above)
Comment (laurit) [reply profiler/src/main/java/com/splunk/opentelemetry/profiler/allocationsampler/SystematicAllocationEventSampler.java:55]: This is tested in `TLABProcessorTest`. Imo not much point in putting too much effort into this as it is throwaway code anyway.

File: profiler/src/main/java/com/splunk/opentelemetry/profiler/allocationsampler/AlwaysOnAllocationEventSampler.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (laurit) [reply profiler/src/main/java/com/splunk/opentelemetry/profiler/allocationsampler/AlwaysOnAllocationEventSampler.java]: removed this class

**Review by breedx-splk — APPROVED** [review]

**Pull request merged**